### PR TITLE
resize instance_MOREPHEUS-8

### PR DESCRIPTION
**Pull request by rahulray19**
resize instance_MOREPHEUS-8
+
Fix resize, stop workload creating extra networks and volumes. Some cleanup

### CODE REVIEW

File: src/main/groovy/com/morpheusdata/xen/XenserverProvisionProvider.groovy [hunk 1]
```diff
@@ -1375,4 +1399,219 @@ class XenserverProvisionProvider extends AbstractProvisionProvider implements Wo
...
+			def modified = false
+			def stopped = false
+			def stopResults
+			def doStop = computeServer.hotResize != true && (neededMemory > 100000000l || neededMemory < -100000000l || neededCores != 0 || resizeRequest.volumesUpdate?.size() > 0)
+			if (doStop) {
+				stopped = true
+				opts.stopped = stopped
+				stopResults = stopWorkload(workload)
+				log.debug("stopResults ${stopResults}")
+			}
```
[the remark this answers is not in the document]
Comment (rahulray19) [reply src/main/groovy/com/morpheusdata/xen/XenserverProvisionProvider.groovy:1429]: As discussed, stop/start workload code not removed.

File: src/main/groovy/com/morpheusdata/xen/XenserverProvisionProvider.groovy [hunk 2]
```diff
@@ -1375,4 +1399,219 @@ class XenserverProvisionProvider extends AbstractProvisionProvider implements Wo
...
+			if (neededMemory > 100000000l || neededMemory < -100000000l || neededCores != 0) {
+				log.info("resizing vm: ${allocationSpecs}")
+				def allocationResults = XenComputeUtility.adjustVmResources(authConfigMap, computeServer.externalId, allocationSpecs)
```
[the remark this answers is not in the document]
Comment (rahulray19) [reply src/main/groovy/com/morpheusdata/xen/XenserverProvisionProvider.groovy:1432]: changes implemented as suggested

File: src/main/groovy/com/morpheusdata/xen/XenserverProvisionProvider.groovy [hunk 3]
```diff
@@ -1375,4 +1399,219 @@ class XenserverProvisionProvider extends AbstractProvisionProvider implements Wo
...
+					workload.maxMemory = requestedMemory.toLong()
+					workload.setConfigProperty('maxCores', (requestedCores ?: 1))
+					workload.maxCores = (requestedCores ?: 1).toLong()
+					workload.plan = plan
+					computeServer.plan = plan
+					computeServer.maxCores = (requestedCores ?: 1).toLong()
+					computeServer.maxMemory = requestedMemory.toLong()
+					context.async.workload.save(workload).blockingGet()
+					computeServer = saveAndGet(computeServer)
+				}
```
[the remark this answers is not in the document]
Comment (rahulray19) [reply src/main/groovy/com/morpheusdata/xen/XenserverProvisionProvider.groovy:1438]: changes added as per suggestion

File: src/main/groovy/com/morpheusdata/xen/XenserverProvisionProvider.groovy [hunk 4]
```diff
@@ -1375,4 +1399,219 @@ class XenserverProvisionProvider extends AbstractProvisionProvider implements Wo
...
+			if (!rtn.error && maxStorage)
+				workload.maxStorage = maxStorage
+			computeServer.status = 'provisioned'
+			computeServer = saveAndGet(computeServer)
+			workload.server = computeServer
+			if (stopped == true) {
+				startWorkload(workload)
+				if (workload.status == Workload.Status.running) {
+				}
+			}
```
[the remark this answers is not in the document]
Comment (rahulray19) [reply src/main/groovy/com/morpheusdata/xen/XenserverProvisionProvider.groovy:1534]: As discussed, start workload code not removed(getting error when removed)

File: src/main/groovy/com/morpheusdata/xen/XenserverProvisionProvider.groovy [hunk 5]
```diff
@@ -1375,4 +1399,219 @@ class XenserverProvisionProvider extends AbstractProvisionProvider implements Wo
...
+					if (deleteResults.success == true) {
+						context.async.computeServer.computeServerInterface.remove([networkDelete], computeServer).blockingGet()
+						computeServer.interfaces.remove(networkDelete)
+						computeServer = saveAndGet(computeServer)
+						workload.server = computeServer
+					}
+				}
+			}
+			workload = context.async.workload.save(workload).blockingGet()
+			if (!rtn.error && maxStorage)
```
[the remark this answers is not in the document]
Comment (rahulray19) [reply src/main/groovy/com/morpheusdata/xen/XenserverProvisionProvider.groovy]: removed this if block as never set in code

File: src/main/groovy/com/morpheusdata/xen/XenserverProvisionProvider.groovy [hunk 6]
```diff
@@ -1375,4 +1399,219 @@ class XenserverProvisionProvider extends AbstractProvisionProvider implements Wo
...
+						def volumeType = storageVolumeTypes[volumeAdd.storageType.toLong()]
+						newVolume.type = volumeType
+						newVolume.datastore = datastore
+						newVolume.uniqueId = "morpheus-vol-${instance.id}-${workload.id}-${newCounter}"
+						computeServer.volumes << newVolume
+						setVolumeInfo(computeServer.volumes, addDiskResults.volumes)
+						context.async.storageVolume.create([newVolume], computeServer).blockingGet()
+						computeServer = saveAndGet(computeServer)
+						workload.server = computeServer
+						context.async.workload.save(workload).blockingGet()
```
[the remark this answers is not in the document]
Comment (rahulray19) [reply src/main/groovy/com/morpheusdata/xen/XenserverProvisionProvider.groovy]: review comment changes incorporated

File: src/main/groovy/com/morpheusdata/xen/XenserverProvisionProvider.groovy [hunk 7]
```diff
@@ -1375,4 +1399,219 @@ class XenserverProvisionProvider extends AbstractProvisionProvider implements Wo
...
+				}
+				resizeRequest.volumesDelete.each { volume ->
+					def deleteResults = XenComputeUtility.deleteVmDisk(authConfigMap, computeServer.externalId, volume.internalId)
+					log.debug("deleteResults ${deleteResults}")
+					if (deleteResults.success == true) {
+						workload = context.async.workload.get(workload.id).blockingGet()
+						computeServer = workload.server
+						computeServer.volumes.remove(volume)
+						context.async.storageVolume.remove([volume], computeServer, true).blockingGet()
+						computeServer = saveAndGet(computeServer)
```
[the remark this answers is not in the document]
Comment (rahulray19) [reply src/main/groovy/com/morpheusdata/xen/XenserverProvisionProvider.groovy]: review comment changes incorporated as suggested

File: src/main/groovy/com/morpheusdata/xen/XenserverProvisionProvider.groovy [hunk 8]
```diff
@@ -1375,4 +1399,219 @@ class XenserverProvisionProvider extends AbstractProvisionProvider implements Wo
...
+					log.debug("netdeleteResults: ${deleteResults}")
+					if (deleteResults.success == true) {
+						context.async.computeServer.computeServerInterface.remove([networkDelete], computeServer).blockingGet()
+						computeServer.interfaces.remove(networkDelete)
+						computeServer = saveAndGet(computeServer)
+						workload.server = computeServer
+					}
+				}
+			}
+			workload = context.async.workload.save(workload).blockingGet()
```
[the remark this answers is not in the document]
Comment (rahulray19) [reply src/main/groovy/com/morpheusdata/xen/XenserverProvisionProvider.groovy]: review comment changes incorporated as suggested

File: src/main/groovy/com/morpheusdata/xen/XenserverProvisionProvider.groovy [hunk 9]
```diff
@@ -1375,4 +1399,219 @@ class XenserverProvisionProvider extends AbstractProvisionProvider implements Wo
...
+					if (addDiskResults.success == true) {
+						def newVolume = buildStorageVolume(computeServer, volumeAdd, addDiskResults, newCounter)
+						def volumeType = storageVolumeTypes[volumeAdd.storageType.toLong()]
+						newVolume.type = volumeType
+						newVolume.datastore = datastore
+						newVolume.uniqueId = "morpheus-vol-${instance.id}-${workload.id}-${newCounter}"
+						computeServer.volumes << newVolume
+						setVolumeInfo(computeServer.volumes, addDiskResults.volumes)
+						context.async.storageVolume.create([newVolume], computeServer).blockingGet()
+						computeServer = saveAndGet(computeServer)
```
[the remark this answers is not in the document]
Comment (rahulray19) [reply src/main/groovy/com/morpheusdata/xen/XenserverProvisionProvider.groovy]: review comment changes incorporated as suggested

File: src/main/groovy/com/morpheusdata/xen/XenserverProvisionProvider.groovy [hunk 10]
```diff
@@ -1375,4 +1399,219 @@ class XenserverProvisionProvider extends AbstractProvisionProvider implements Wo
...
+					addDiskConfig.datastoreId = datastore?.externalId
+					def addDiskResults = XenComputeUtility.addVmDisk(authConfigMap, computeServer.externalId, addDiskConfig)
+					log.debug("addDiskResults ${addDiskResults}")
+					if (addDiskResults.success == true) {
+						def newVolume = buildStorageVolume(computeServer, volumeAdd, addDiskResults, newCounter)
+						def volumeType = storageVolumeTypes[volumeAdd.storageType.toLong()]
+						newVolume.type = volumeType
+						newVolume.datastore = datastore
+						newVolume.uniqueId = "morpheus-vol-${instance.id}-${workload.id}-${newCounter}"
+						computeServer.volumes << newVolume
```
[the remark this answers is not in the document]
Comment (rahulray19) [reply src/main/groovy/com/morpheusdata/xen/XenserverProvisionProvider.groovy]: review comment changes incorporated as suggested

File: src/main/groovy/com/morpheusdata/xen/XenserverProvisionProvider.groovy [hunk 11]
```diff
@@ -1375,4 +1399,219 @@ class XenserverProvisionProvider extends AbstractProvisionProvider implements Wo
...
+					} else {
+						log.warn("error adding disk: ${addDiskResults}")
+					}
+				}
+				resizeRequest.volumesDelete.each { volume ->
+					def deleteResults = XenComputeUtility.deleteVmDisk(authConfigMap, computeServer.externalId, volume.internalId)
+					log.debug("deleteResults ${deleteResults}")
+					if (deleteResults.success == true) {
+						workload = context.async.workload.get(workload.id).blockingGet()
+						computeServer = workload.server
```
[the remark this answers is not in the document]
Comment (rahulray19) [reply src/main/groovy/com/morpheusdata/xen/XenserverProvisionProvider.groovy]: review comment changes incorporated as suggested

File: src/main/groovy/com/morpheusdata/xen/XenserverProvisionProvider.groovy [hunk 12]
```diff
@@ -1375,4 +1399,219 @@ class XenserverProvisionProvider extends AbstractProvisionProvider implements Wo
...
+						computeServer.volumes.remove(volume)
+						context.async.storageVolume.remove([volume], computeServer, true).blockingGet()
+						computeServer = saveAndGet(computeServer)
+						workload.server = computeServer
+						context.async.workload.save(workload).blockingGet()
```
[the remark this answers is not in the document]
Comment (rahulray19) [reply src/main/groovy/com/morpheusdata/xen/XenserverProvisionProvider.groovy]: review comment changes incorporated as suggested

File: src/main/groovy/com/morpheusdata/xen/XenserverProvisionProvider.groovy [hunk 13]
```diff
@@ -1375,4 +1399,219 @@ class XenserverProvisionProvider extends AbstractProvisionProvider implements Wo
...
+					def networkConfig = [networkIndex: newIndex, networkUuid: newNetwork.externalId]
+					def networkResults = XenComputeUtility.addVmNetwork(authConfigMap, computeServer.externalId, networkConfig)
+					log.debug("networkResults ${networkResults}")
+					if (networkResults.success == true) {
+						def newInterface = buildNetworkInterface(computeServer, networkResults, newNetwork, newIndex, index)
+						newInterface.uniqueId = "morpheus-nic-${instance.id}-${workload.id}-${newIndex}"
+						computeServer.interfaces << newInterface
+						context.async.computeServer.computeServerInterface.create([newInterface], computeServer).blockingGet()
+						computeServer = saveAndGet(computeServer)
+						workload.server = computeServer
```
[the remark this answers is not in the document]
Comment (rahulray19) [reply src/main/groovy/com/morpheusdata/xen/XenserverProvisionProvider.groovy]: review comment changes incorporated as suggested

File: src/main/groovy/com/morpheusdata/xen/XenserverProvisionProvider.groovy [hunk 14]
```diff
@@ -1375,4 +1399,219 @@ class XenserverProvisionProvider extends AbstractProvisionProvider implements Wo
...
+				}
+				resizeRequest?.interfacesDelete?.eachWithIndex { networkDelete, index ->
+					authConfigMap.stopped = opts.stopped
+					def deleteResults = XenComputeUtility.deleteVmNetwork(authConfigMap, computeServer.externalId, networkDelete.internalId)
+					log.debug("netdeleteResults: ${deleteResults}")
+					if (deleteResults.success == true) {
+						context.async.computeServer.computeServerInterface.remove([networkDelete], computeServer).blockingGet()
+						computeServer.interfaces.remove(networkDelete)
+						computeServer = saveAndGet(computeServer)
+						workload.server = computeServer
```
[the remark this answers is not in the document]
Comment (rahulray19) [reply src/main/groovy/com/morpheusdata/xen/XenserverProvisionProvider.groovy]: review comment changes incorporated as suggested

File: src/main/groovy/com/morpheusdata/xen/XenserverProvisionProvider.groovy [hunk 15]
```diff
@@ -1375,4 +1399,219 @@ class XenserverProvisionProvider extends AbstractProvisionProvider implements Wo
...
+					}
+				}
+			}
+			workload = context.async.workload.save(workload).blockingGet()
+			if (!rtn.error && maxStorage)
+				workload.maxStorage = maxStorage
+			computeServer.status = 'provisioned'
+			computeServer = saveAndGet(computeServer)
+			workload.server = computeServer
```
[the remark this answers is not in the document]
Comment (rahulray19) [reply src/main/groovy/com/morpheusdata/xen/XenserverProvisionProvider.groovy]: review comment changes incorporated as suggested

File: src/main/groovy/com/morpheusdata/xen/XenserverProvisionProvider.groovy [hunk 16]
```diff
@@ -1375,4 +1399,219 @@ class XenserverProvisionProvider extends AbstractProvisionProvider implements Wo
...
+			if (stopped == true) {
+				startWorkload(workload)
+				if (workload.status == Workload.Status.running) {
+				}
+			}
+			rtn.success = true
+		} catch (e) {
+			log.error("Unable to resize workload: ${e.message}", e)
+			workload.server.status = 'provisioned'
+			workload.server = saveAndGet(workload.server)
```
[the remark this answers is not in the document]
Comment (rahulray19) [reply src/main/groovy/com/morpheusdata/xen/XenserverProvisionProvider.groovy]: review comment changes incorporated as suggested

File: src/main/groovy/com/morpheusdata/xen/XenserverProvisionProvider.groovy [hunk 17]
```diff
@@ -1375,4 +1399,219 @@ class XenserverProvisionProvider extends AbstractProvisionProvider implements Wo
...
+		ServiceResponse rtn = ServiceResponse.success()
+		ComputeServer computeServer = context.async.computeServer.get(workload.server.id).blockingGet()
+		Cloud cloud = computeServer.cloud
+		def authConfigMap = plugin.getAuthConfig(cloud)
+		try {
+			computeServer.status = 'resizing'
+			computeServer = saveAndGet(computeServer)
+			ServicePlan plan = resizeRequest.plan
+			def requestedMemory = plan.customMaxMemory ? (resizeRequest.maxMemory ?: plan.maxMemory) : plan.maxMemory
+			def requestedCores = plan.customCores ? (resizeRequest?.maxCores ?: plan.maxCores) : plan.maxCores
```
[the remark this answers is not in the document]
Comment (rahulray19) [reply src/main/groovy/com/morpheusdata/xen/XenserverProvisionProvider.groovy]: review comment changes incorporated as suggested

**Review by ddeyoung — APPROVED** [review]

**Pull request merged**